### PR TITLE
INT: Support "Invert if condition" without else branch

### PR DESCRIPTION
**Pull request by dima74**
output

changelog: Support "Invert if condition" without else branch

### CODE REVIEW

**Review by ortem — CHANGES_REQUESTED** [review]
LGTM, I tested the implementation on some real code, and it seems to work very well!
Except for my (mostly minor) comments, I think it's worth describing the implementation in more detail since the logic is quite difficult to understand at first sight. So it would be nice to have some comments inside the code

File: src/main/kotlin/org/rust/ide/intentions/InvertIfIntention.kt [hunk 1]
```diff
@@ -7,18 +7,45 @@ package org.rust.ide.intentions
...
+    class ContextWithoutElse(
+        val ifExpr: RsIfExpr,
+        override val ifCondition: RsExpr,
+        /**
+         * - [ifExpr] if it is tail expr
+         * - `ifExpr.parent` if stmt
+         */
+        val ifStmt: RsElement,
+        val ifBlock: RsBlock,
+        val ifStmts: List<PsiElement>,
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/ide/intentions/InvertIfIntention.kt]: I would keep `thenBlockStmts` name — there also can be whitespaces/comments, etc, but the core part of this list are statements

File: src/main/kotlin/org/rust/ide/intentions/InvertIfIntention.kt [hunk 2]
```diff
@@ -27,29 +88,254 @@ class InvertIfIntention : RsElementBaseIntentionAction<InvertIfIntention.Context
...
+    any { it !is PsiWhiteSpace && it !is PsiComment }
+
+private fun RsExpr.canBeStmtWithoutSemicolon(): Boolean =
+    this is RsWhileExpr || this is RsForExpr || this is RsLoopExpr
+        || this is RsIfExpr || this is RsMatchExpr || this is RsBlockExpr
+
+private fun RsExpr.wrapInStmt(factory: RsPsiFactory): RsExprStmt =
+    factory.tryCreateExprStmt(text)!!
+
+private fun PsiElement.isLoop(): Boolean =
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/intentions/InvertIfIntention.kt:323]: [minor] Can be replaced by `is RsLooplikeExpr` check

**Review by ortem — APPROVED** [review]
LGTM except for one minor comment

**Comment by dima74** [comment]
bors r=ortem

**Pull request merged**